src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java

```diff
@@ -0,0 +1,422 @@
+                    }
+                }
+                if (portStr == null) {
+                    firstError = chain(firstError, new IOException(jenkinsUrl + " is not Jenkins"));
+                    continue;
+                }
+                int port = 0;
+                try {
+                    port = Integer.parseInt(portStr);
+                } catch (NumberFormatException e) {
+                    firstError = chain(firstError, new IOException(jenkinsUrl + " is publishing an invalid port", e));
+                    continue;
+                }
+                if (port <= 0 || 65536 <= port) {
+                    firstError = chain(firstError, new IOException(jenkinsUrl + " is publishing an invalid port"));
+                    continue;
+                }
+                // sort the URLs so that the winner is the one we try first next time
+                final String winningJenkinsUrl = jenkinsUrl;
+                Collections.sort(jenkinsUrls, new Comparator<String>() {
+                    @Override
+                    public int compare(String o1, String o2) {
+                        if (winningJenkinsUrl.equals(o1)) {
+                            return -1;
+                        }
+                        if (winningJenkinsUrl.equals(o2)) {
+                            return 1;
+                        }
+                        return 0;
+                    }
+                });
+                if (tunnel != null) {
+                    String[] tokens = tunnel.split(":", 3);
+                    if (tokens.length != 2)
+                        throw new IOException("Illegal tunneling parameter: " + tunnel);
+                    if (tokens[0].length() > 0) host = tokens[0];
+                    if (tokens[1].length() > 0) port = Integer.parseInt(tokens[1]);
+                }
+                return new JnlpAgentEndpoint(host, port, identity, agentProtocolNames);
+            } finally {
+                con.disconnect();
+            }
+        }
+        if (firstError != null) {
+            throw firstError;
+        }
+        return null;
+    }
+
+    private URL toAgentListenerURL(String jenkinsUrl) throws MalformedURLException {
+        return jenkinsUrl.endsWith("/")
+                ? new URL(jenkinsUrl + "tcpSlaveAgentListener/")
+                : new URL(jenkinsUrl + "/tcpSlaveAgentListener/");
+    }
+
+    public void waitForReady() throws InterruptedException {
+        Thread t = Thread.currentThread();
+        String oldName = t.getName();
+        try {
+            int retries = 0;
+            while (true) {
+                Thread.sleep(1000 * 10);
+                try {
+                    // Jenkins top page might be read-protected. see http://www.nabble
+                    // .com/more-lenient-retry-logic-in-Engine.waitForServerToBack-td24703172.html
+                    URL url = toAgentListenerURL(first(jenkinsUrls));
```

Comment:
maybe we should try different jenkins URLs in the loop since we have this loop in any case